### PR TITLE
HTTP: Implement persistent connections (keep-alive) etc.

### PR DESCRIPTION
**Pull request by DDvO**
This PR implements feature request #14004.

The proposed semantics of the new feature implemented here is as follows.
Both at API and at CLI level (for the CMP app only, so far) there is a new parameter/option: `keep_alive`. 
* 0 means that HTTP connections are not kept open after receiving a response, which is the default behavior for HTTP 1.0.
* 1 means that persistent connections are requested.
* 2 means that persistent connections are required, i.e., in case the server does not grant them an error occurs.

For the CMP app the default value is 1, which means preferring to keep the connection open.
For all other internal uses of the HTTP client (fetching an OCSP response, a cert, or a CRL) it does not matter because these operations just take one round trip.

If the client application requested or required a persistent connection and this was granted by the server,
 it can keep the `OSSL_HTTP_REQ_CTX *` as long as it wants to send further requests and `OSSL_HTTP_is_alive()` returns nonzero,
else it should call `OSSL_HTTP_REQ_CTX_free()` or `OSSL_HTTP_close()`.
In case the client application keeps the `OSSL_HTTP_REQ_CTX *` but the connection then dies for any reason at the server side, it will notice this obtaining an I/O error when trying to send the next request .

This feature addition requires extending the HTTP header parsing and rearranging the high-level HTTP client API. 
In particular:
* Split the monolithic `OSSL_HTTP_transfer()` into `OSSL_HTTP_open()`,
  `OSSL_HTTP_set_request(),` a lean `OSSL_HTTP_transfer(),` and `OSSL_HTTP_close().`
* Split the timeout functionality accordingly and improve default behavior.
* Extract part of `OSSL_HTTP_REQ_CTX_new()` to `OSSL_HTTP_REQ_CTX_set_expected().`
* Extend the local `struct ossl_http_req_ctx_st` accordingly.

Also extend related diagnostic output for debugging.
Use the new feature for the CMP client, 
which requires extending related transaction management of CMP client and test server.

Related further more general improvements (in separate commits) included because they are related:
* `OSSL_HTTP_get():` Do not close connection if redirect to same server
* `OSSL_HTTP_REQ_CTX_add1_headers():` Fix use with `host == NULL` (relative URLs)
* `OSSL_HTTP_transfer()`: Fix error reporting in case `rctx->server == NULL`.
    Also improve doc of `OSSL_parse_url()` and `OSSL_HTTP_parse_url()`
* Generalize API to arbitrary request and response contents 
  ~Introducing `OSSL_HTTP_i2d()` and `OSSL_HTTP_d2i()` makes the API more clear and modular.~
* Allow streaming of request data (for POST method).
  Also clean up `OSSL_HTTP_REQ_CTX_nbio`() states and make it more efficient.
* Allow streaming of response data (with possibly indefinite length).
   Also clean up `max_resp_len` and add `OSSL_HTTP_REQ_CTX_get_resp_len()`
* Rename `maxline` parameter to `buf_size` for clarity.
* Rename internal fields and functions in `http_client.c` for consistency.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
When submitting this PR some 2 days ago, I forgot to mention details about the proposed semantics of the new feature. Here they are:

* If `keep_alive` is 0 then HTTP connections are not kept open after receiving a response, which is the default behavior for HTTP 1.0.
* If the value is 1 or 2 then persistent connections are requested.
* If the value is 2 then persistent connections are required,
i.e., in case the server does not grant them an error occurs.

For the CMP app the default value is 1, which means preferring to keep the connection open.
For all other internal uses of the HTTP client (fetching an OCSP response, a cert, or a CRL) it does not matter because these operations just take one round trip.

If the client application requested or required a persistent connection and this was granted by the server,
 it can keep the `OSSL_HTTP_REQ_CTX *` as long as it wants to send further requests and `OSSL_HTTP_REQ_CTX_is_alive()` returns nonzero,
else it should call `OSSL_HTTP_REQ_CTX_free()` or `OSSL_HTTP_close()`.
In case the client application keeps the `OSSL_HTTP_REQ_CTX *` but the connection then dies for any reason at the server side, it will notice this obtaining an I/O error when trying to send the next request .

**Comment by DDvO** [comment]
@openssl/otc, is there already someone willing to review this?
This must get in before feature freeze, also due to the API changes.

**Comment by DDvO** [comment]
BTW, I also have some HTTP client API streamlining and generalization in the pipeline to be submitted as follow-up;
would be good if this will make it into 3.0 as well...

**Comment by DDvO** [comment]
I meanwhile found that specific tests for the new keep-alive feature can be added in a rather straightforward way, so done this as well.

**Comment by DDvO** [comment]
Also added a couple of related small fixes of the HTTP client implementation,
mostly regarding the special case of relative URLs, which came up due to the new tests.

**Comment by DDvO** [comment]
Added a fixup on the documentation of `OSSL_HTTP_REQ_CTX_set_expected()`,
more clearly explaining the default behavior if this new low-level HTTP client API function is not used.

**Comment by DDvO** [comment]
Since so far I have not seen any reaction on this PR that would indicate that someone has started reviewing this, 
I've used the time to further generalize and actually simplify the high-level HTTP client API by making the `OSSL_HTTP_{get,transfer}_asn1()` functions independent of the request and response content type.

**Comment by DDvO** [comment]
On this occasion I've updated the `NEWS.md` and
extended the `CHANGES.md` entry on the HTTP client as follows:
```
   It supports arbitrary request and response content types, GET redirection,
   TLS, connections via HTTP(S) proxies, connections and exchange via
   user-defined BIOs (allowing implicit connections), persistent connections,
   and timeout checks.  See L<OSSL_HTTP_transfer(3)> etc. for details.
   The legacy OCSP-focused (and only partly documented) API
   is retained for backward compatibility, while most of it is deprecated.
 ```

**Comment by DDvO** [comment]
@openssl/otc, **I am getting more and more concerned** about how to get this PR approved and merged before the public API / feature freeze, which is going to be in 16 days from now.

This PR, submitted 6 days ago, implements a feature discussed in #14004. It is critical for one of our customers in order to be able to make their product compliant with ERTMS/ECTS requirements by UNISIG.
Unfortunately, so far I did not receive any reaction on it, while OTOH the deadline for getting public API changes included into 3.0 is approaching rapidly.

How to solve this?
Would it make sense for instance if I carve out the pure API changes such that they are reviewed and approved hopefully quicker and are safely merged before the deadline,
and then add the internal feature implementation as well as the related tests and documentation later (hopefully already into beta1, otherwise during the beta phase)?

**Comment by DDvO** [comment]
Rebased again to fix simple merge conflicts

File: apps/lib/http_server.c [hunk 1]
```diff
@@ -429,21 +497,27 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
 
 /* assumes that cbio does not do an encoding that changes the output length */
-int http_server_send_asn1_resp(BIO *cbio, const char *content_type,
+int http_server_send_asn1_resp(BIO *cbio, int keep_alive,
+                               const char *content_type,
                                const ASN1_ITEM *it, const ASN1_VALUE *resp)
 {
-    int ret = BIO_printf(cbio, "HTTP/1.0 200 OK\r\nContent-type: %s\r\n"
-                         "Content-Length: %d\r\n\r\n", content_type,
+    int ret = BIO_printf(cbio, HTTP_PREFIX_VERSION"0 200 OK\r\n%s" /* HTTP/1.0*/
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:504]: nit: `/* HTTP/1.0 */`

File: apps/lib/http_server.c [hunk 2]
```diff
@@ -429,21 +497,27 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
+                         "Content-type: %s\r\n"
+                         "Content-Length: %d\r\n\r\n",
+                         keep_alive ? "Connection: keep-alive\r\n" : "",
+                         content_type,
                          ASN1_item_i2d(resp, NULL, it)) > 0
             && ASN1_item_i2d_bio(it, cbio, resp) > 0;
-
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:439]: nit keep blank line after decl.

File: apps/lib/http_server.c [hunk 3]
```diff
@@ -429,21 +497,27 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
     (void)BIO_flush(cbio);
     return ret;
 }
 
 int http_server_send_status(BIO *cbio, int status, const char *reason)
 {
-    int ret = BIO_printf(cbio, "HTTP/1.0 %d %s\r\n\r\n", status, reason) > 0;
+    int ret = BIO_printf(cbio,
+                         HTTP_PREFIX_VERSION"0 %d %s\r\n\r\n", /* HTTP/1.0*/
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:518]: nit: same `.../1.0 */`

File: doc/man3/OSSL_CMP_SRV_CTX_new.pod [hunk 4]
```diff
@@ -89,6 +89,10 @@ Its arguments are the B<OSSL_CMP_SRV_CTX> I<srv_ctx> and the CMP request message
 I<req>. It does the typical generic checks on I<req>, calls
 the respective callback function (if present) for more specific processing,
 and then assembles a result message, which may be a CMP error message.
+If after return of the function
```
Comment (FdaSilvaYY) [inline doc/man3/OSSL_CMP_SRV_CTX_new.pod:92]: nit: if return result of I<function> is ...

File: include/openssl/http.h [hunk 5]
```diff
@@ -38,63 +36,64 @@ typedef BIO *(*OSSL_HTTP_bio_cb_t)(BIO *bio, void *arg, int connect, int detail)
...
                         OSSL_HTTP_bio_cb_t bio_update_fn, void *arg,
-                        const STACK_OF(CONF_VALUE) *headers,
-                        const char *content_type, BIO *req_mem,
-                        int maxline, unsigned long max_resp_len, int timeout,
-                        const char *expected_ct, int expect_asn1,
-                        char **redirection_url);
-int OSSL_HTTP_proxy_connect(BIO *bio, const char *server, const char *port,
-                            const char *proxyuser, const char *proxypass,
-                            int timeout, BIO *bio_err, const char *prog);
+                        int buf_size,const STACK_OF(CONF_VALUE) *headers,
```
Comment (FdaSilvaYY) [inline include/openssl/http.h:90]: nit: missing after comma

File: test/http_test.c [hunk 6]
```diff
@@ -17,43 +17,43 @@
...
+#define RPATH "/path/result.crt"
+
+typedef struct {
+    BIO *out;
+    char version;
+    int keep_alive;
+} server_args;
+
+/*-
+ * Ppretty trivial HTTP mock server:
```
Comment (FdaSilvaYY) [inline test/http_test.c:29]: typo: Ppretty

File: doc/man3/OSSL_HTTP_transfer.pod [hunk 7]
```diff
@@ -200,23 +228,28 @@ other HTTP client implementations such as wget, curl, and git.
...
-On success, OSSL_HTTP_get(), OSSL_HTTP_get_asn1(), OSSL_HTTP_post_asn1(), and
-OSSL_HTTP_transfer() return a memory BIO containing the data received via HTTP.
-This must be freed by the caller. On failure, NULL is returned.
+OSSL_HTTP_open() returns on success a B<OSSL_HTTP_REQ_CTX>, else NULL.
+
+OSSL_HTTP_proxy_connect() and OSSL_HTTP_set_request()
+return 1 on success, 0 on error.
+
+On success, OSSL_HTTP_exchange(), OSSL_HTTP_get(), and OSSL_HTTP_transfer()
+return a memory BIO containing the data received if asn ASN.1 responsed is
```
Comment (FdaSilvaYY) [inline doc/man3/OSSL_HTTP_transfer.pod:237]: typos: if asn ASN.1 responsed

**Review by FdaSilvaYY — COMMENTED** [review]
great work

File: doc/man3/OSSL_CMP_SRV_CTX_new.pod (hunk 4 above)
Comment (DDvO) [reply doc/man3/OSSL_CMP_SRV_CTX_new.pod]: > nit: if return result of `I<function>` is ...

This was meant differently; extended the sentence for clarity.

**Comment by DDvO** [comment]
> great work

Pleased to hear!

**Comment by DDvO** [comment]
Thanks @FdaSilvaYY for reporting those nits; fixed them.

**Comment by DDvO** [comment]
Carved out an "API-only" PR: #15147 - does that help reviewing?

**Review by t8m — COMMENTED** [review]
Just a few preliminary comments

File: apps/include/http_server.h [hunk 8]
```diff
@@ -93,7 +98,8 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
  * resp: the response to send
  * returns 1 on success, 0 on failure
  */
-int http_server_send_asn1_resp(BIO *cbio, const char *content_type,
+int http_server_send_asn1_resp(BIO *cbio, int keep_alive,
```
[the remark this answers is not in the document]
Comment (DDvO) [reply apps/include/http_server.h:102]: Thanks for pointing this out; done.

File: apps/include/http_server.h [hunk 9]
```diff
@@ -81,7 +84,9 @@ BIO *http_server_init_bio(const char *prog, const char *port);
  */
 int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
                              char **ppath, BIO **pcbio, BIO *acbio,
-                             const char *prog, int accept_get, int timeout);
+                             int *found_keep_alive,
```
[the remark this answers is not in the document]
Comment (DDvO) [reply apps/include/http_server.h:88]: Yeah; fixed as well.

File: apps/cmp.c [hunk 10]
```diff
@@ -2666,7 +2687,9 @@ int cmp_main(int argc, char **argv)
 #else
         BIO *acbio;
         BIO *cbio = NULL;
+        int keep_alive;
```
[the remark this answers is not in the document]
Comment (DDvO) [reply apps/cmp.c]: Done (though not strictly needed - the variable is assigned by `http_server_get_asn1_req()`,
but static checker may have trouble detecting this).

**Comment by DDvO** [comment]
If it helps, once this PR has been rebased on #15147 (when that has been merged)
I could further split this PR into
* the client-side changes
* the changes to the test server and the new tests for the client

**Comment by DDvO** [comment]
Rebased, which was not a good idea at the moment. 
The new check_update CI failure is unrelated.

**Comment by DDvO** [comment]
In order to get rid of the unrelated CI failure on `check_update`
I had to add an artificial commit for updating the FIPS checksums
since apparently this meanwhile required and someone forgot to do so.

**Comment by DDvO** [comment]
Rebased; prepared for rebasing on meanwhile approved #15147 (which extracted the API changes).

**Comment by DDvO** [comment]
@t8m, finally I've finished rebasing this PR on the meanwhile merged #15147,
which was much more work than expected due to many non-trivial conflicts.

So this PR is ready for review again.

**Comment by DDvO** [comment]
BTW, the number of lines changed as reported above but GitHub: `+1,480 −878` is misleading
because there are some intermediate changes among the included commits.
Yet `git diff upstream/master --stat` correctly reports
```
 NEWS.md                                                |   6 +-
 apps/cmp.c                                             |  73 +++++--
 apps/cmp_mock_srv.c                                    |   3 +-
 apps/include/apps.h                                    |   1 +
 apps/include/http_server.h                             |  23 ++-
 apps/lib/apps.c                                        |   7 +-
 apps/lib/http_server.c                                 | 126 +++++++++---
 apps/ocsp.c                                            |  14 +-
 crypto/cmp/cmp_ctx.c                                   |  17 +-
 crypto/cmp/cmp_http.c                                  |  37 +++-
 crypto/cmp/cmp_local.h                                 |   2 +
 crypto/cmp/cmp_server.c                                |  25 ++-
 crypto/err/openssl.txt                                 |   2 +
 crypto/http/http_client.c                              | 663 ++++++++++++++++++++++++++++++++++-------------------------
 crypto/http/http_err.c                                 |   4 +
 crypto/ocsp/ocsp_http.c                                |  24 ++-
 crypto/x509/x_all.c                                    |   2 +-
 doc/man3/OSSL_CMP_CTX_new.pod                          |  13 +-
 doc/man3/OSSL_CMP_SRV_CTX_new.pod                      |   4 +
 doc/man3/OSSL_HTTP_REQ_CTX.pod                         |   6 +-
 doc/man3/OSSL_HTTP_parse_url.pod                       |   8 +-
 doc/man3/OSSL_HTTP_transfer.pod                        |   3 +-
 include/openssl/http.h                                 |   4 +-
 test/http_test.c                                       | 187 +++++++++++++----
 test/recipes/80-test_cmp_http_data/test_connection.csv |  92 +++++----
 25 files changed, 883 insertions(+), 463 deletions(-)
```

**Comment by DDvO** [comment]
@t8m, did you notice that #15221 comprises of four initial commits of this PR?
I thought that providing them could ease reviewing.

**Comment by DDvO** [comment]
I could try further splitting up this PR to help reviewing,
but on the other hand this can cause quite some extra effort 
and would certainly prolong the overall process,
and we are running out of time since the whole thing should be merged within a week from now :frowning_face:

**Comment by slontis** [comment]
Or not merged until later :)

**Comment by DDvO** [comment]
> Or not merged until later :)

I did not get what you mean by this?

**Comment by DDvO** [comment]
BTW, what is the exact terms of the feature/API freeze planned for May 20th?
No changes to any file in `include/openssl`?

Can additions to implementations, apps, documentation, and tests still be done thereafter?

**Comment by mattcaswell** [comment]
> Can additions to implementations (including internal headers), apps, documentation, and tests still be done thereafter?

Missing/incorrect documentation is considered a bug so that is fine. Adding new tests is also fine. Changes to internal headers (e.g. to fix a bug) are ok.

New *features* are not ok. So adding new capabilities to apps etc would not be allowed.

**Comment by DDvO** [comment]
> New features are not ok. So adding new capabilities to apps etc would not be allowed.

Thanks for the clarification.
So most of this PR is **urgent** to handle before the deadline.

**Comment by DDvO** [comment]
Would it help for now if I remove the test additions and minor bug fixes from this PR?
They make up roughly 20% of the total number of changed lines.

File: apps/lib/http_server.c [hunk 11]
```diff
@@ -294,23 +313,32 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
+    if ((end = strchr(reqbuf, '\r')) != NULL
+            || (end = strchr(reqbuf, '\n')) != NULL)
+        *end = '\0';
+    log_message(prog, LOG_INFO, "Received request, 1st line: %s", reqbuf);
 
     meth = reqbuf;
     url = meth + 3;
     if ((accept_get && strncmp(meth, "GET ", 4) == 0)
             || (url++, strncmp(meth, "POST ", 5) == 0)) {
+        const char *const http_version_str = " "HTTP_PREFIX_VERSION;
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:333]: Previously suggested in another PR:
```
        static const char http_version_str[] = " "HTTP_PREFIX_VERSION;
```

File: apps/lib/http_server.c [hunk 12]
```diff
@@ -377,15 +409,50 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
         len = BIO_gets(cbio, inbuf, sizeof(inbuf));
         if (len <= 0) {
-            log_message(prog, LOG_ERR,
-                        "Error skipping remaining HTTP headers");
+            log_message(prog, LOG_WARNING, "Error reading HTTP header");
             (void)http_server_send_status(cbio, 400, "Bad Request");
             goto out;
         }
+
         if ((inbuf[0] == '\r') || (inbuf[0] == '\n'))
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:421]: nit : useless parenthesis

File: apps/lib/http_server.c [hunk 13]
```diff
@@ -429,11 +498,15 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
...
-int http_server_send_asn1_resp(BIO *cbio, const char *content_type,
+int http_server_send_asn1_resp(BIO *cbio, int keep_alive,
+                               const char *content_type,
                                const ASN1_ITEM *it, const ASN1_VALUE *resp)
 {
-    int ret = BIO_printf(cbio, "HTTP/1.0 200 OK\r\nContent-type: %s\r\n"
-                         "Content-Length: %d\r\n\r\n", content_type,
+    int ret = BIO_printf(cbio, HTTP_PREFIX_VERSION"0 200 OK\r\n%s" /* HTTP/1.0 */
+                         "Content-type: %s\r\n"
+                         "Content-Length: %d\r\n\r\n",
```
Comment (FdaSilvaYY) [inline apps/lib/http_server.c:507]: "Content-Length: %d" will result in outputting "Content-Length: 0" or "Content-Length: 1", here ?
Disclaimer: I'm not HTTP fluent ;)

File: crypto/cmp/cmp_http.c [hunk 14]
```diff
@@ -55,24 +55,45 @@ OSSL_CMP_MSG *OSSL_CMP_MSG_http_perform(OSSL_CMP_CTX *ctx,
...
-    ossl_cmp_log2(DEBUG, ctx, "connecting to CMP server %s%s",
-                  ctx->server, tls_used ? " using TLS" : "");
-    rsp = OSSL_HTTP_transfer(NULL, ctx->server, server_port,
+    if (ctx->http_ctx == NULL)
+        ossl_cmp_log3(DEBUG, ctx, "connecting to CMP server %s:%s%s",
+                      ctx->server, server_port, tls_used ? " using TLS" : "");
+
+    type = req->body->type;
+    keep_alive = ctx->keep_alive;
+    if (keep_alive != 0
```
Comment (FdaSilvaYY) [inline crypto/cmp/cmp_http.c:65]: Suggestion : extract this `if` into a separate method. like 
```
/* really enable keep alive depending on request type */
keep_alive = should_keep_alive(ctx->keep_alive, req->body->type);
```
you will spare the `type` local variable

File: crypto/http/http_client.c [hunk 15]
```diff
@@ -175,9 +198,10 @@ int OSSL_HTTP_REQ_CTX_set_request_line(OSSL_HTTP_REQ_CTX *rctx, int method_POST,
     if (path[0] != '/' && BIO_printf(rctx->mem, "/") <= 0)
         return 0;
 
-    if (BIO_printf(rctx->mem, "%s "HTTP_PREFIX"1.0\r\n", path) <= 0)
+    if (BIO_printf(rctx->mem, "%s "HTTP_PREFIX_VERSION"0\r\n", path) <= 0)
```
Comment (FdaSilvaYY) [inline crypto/http/http_client.c:201]: Better comment on end of line, the real value : `/* HTTP/1.0 */`
As it was done in other places

File: apps/lib/http_server.c (hunk 11 above)
Comment (DDvO) [reply apps/lib/http_server.c]: Done - just cannot use `static` due to the subsequent line:
```
size_t http_version_str_len = strlen(http_version_str)
```
which the compiler cannot handle statically (even if I declare `http_version_str_len` static, too).

File: apps/lib/http_server.c (hunk 12 above)
Comment (DDvO) [reply apps/lib/http_server.c]: Removed

File: apps/lib/http_server.c (hunk 13 above)
Comment (DDvO) [reply apps/lib/http_server.c:508]: You just need to be C fluent here :)
Note that the parenthesis closes before the comparison operator:
```
ASN1_item_i2d(resp, NULL, it)) > 0
```

File: crypto/cmp/cmp_http.c (hunk 14 above)
Comment (DDvO) [reply crypto/cmp/cmp_http.c]: Good thought :+1:
Done

File: crypto/http/http_client.c (hunk 15 above)
Comment (DDvO) [reply crypto/http/http_client.c]: Yeah, I think it's even better to factor out all these uses - done as follows:
```
#define HTTP_PREFIX_V1_0 HTTP_PREFIX_VERSION"0" /* "HTTP/1.0" */
```

**Comment by DDvO** [comment]
Thanks @FdaSilvaYY for your comments - handled all of them.

**Comment by DDvO** [comment]
The new CI failure is unrelated.

**Comment by t8m** [comment]
Hmm, the fips checksum labeling does not work well if the branch is not freshly rebased.

**Comment by DDvO** [comment]
> Hmm, the fips checksum labeling does not work well if the branch is not freshly rebased.

Rebased (without squashing recent fixups).

**Comment by DDvO** [comment]
Can I somehow rebase this PR already now (before merging to master in 24 hours) on 15221?

File: apps/lib/http_server.c (hunk 11 above)
Comment (t8m) [reply apps/lib/http_server.c]: You could use `http_version_str_len = sizeof(http_version_str) - 1;` if http_version_str is array.

**Comment by t8m** [comment]
As the #15221 will be changed when merging due to the commit annotations, there is no point.

**Comment by DDvO** [comment]
I see.
Is there anything else I can do now to ease reviewing this PR?

File: apps/lib/http_server.c (hunk 11 above)
Comment (DDvO) [reply apps/lib/http_server.c]: Good thought - done.

File: include/openssl/http.h [hunk 16]
```diff
@@ -33,8 +33,8 @@ extern "C" {
 # define OPENSSL_HTTP_PROXY "HTTP_PROXY"
 # define OPENSSL_HTTPS_PROXY "HTTPS_PROXY"
 
-#define HTTP_DEFAULT_MAX_LINE_LENGTH (4 * 1024)
-#define HTTP_DEFAULT_MAX_RESP_LEN (100 * 1024)
+#define OSSL_HTTP_DEFAULT_MAX_LINE_LENGTH (4 * 1024)
```
Comment (t8m) [inline include/openssl/http.h:36]: Why is one named with _LENGTH and the other with _LEN suffix? Could you unify to _LEN as the names are long enough?

**Review by t8m — CHANGES_REQUESTED** [review]
Just nits. Otherwise great work!

File: crypto/http/http_client.c [hunk 17]
```diff
@@ -24,57 +24,71 @@
 #include "internal/cryptlib.h" /* for ossl_assert() */
 
 #include "http_local.h"
-
```
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/http/http_client.c:27]: This must have been due to a glitch when solving merge conflicts or the like. Reverted.

File: apps/lib/http_server.c [hunk 18]
```diff
@@ -330,8 +331,8 @@ int http_server_get_asn1_req(const ASN1_ITEM *it, ASN1_VALUE **preq,
     url = meth + 3;
     if ((accept_get && strncmp(meth, "GET ", 4) == 0)
             || (url++, strncmp(meth, "POST ", 5) == 0)) {
-        const char *const http_version_str = " "HTTP_PREFIX_VERSION;
-        size_t http_version_str_len = strlen(http_version_str); /* " HTTP/1." */
+        static const char http_version_str[] = " "HTTP_PREFIX_VERSION;
+        static size_t http_version_str_len = sizeof(http_version_str) - 1;
```
[the remark this answers is not in the document]
Comment (DDvO) [reply apps/lib/http_server.c]: Done.

File: crypto/http/http_client.c [hunk 19]
```diff
@@ -256,12 +284,10 @@ static int ossl_http_req_ctx_set_content(OSSL_HTTP_REQ_CTX *rctx,
...
         return 0;
 
-    if ((req_len = BIO_get_mem_data(req_mem, &req)) <= 0)
-        return 0;
-    rctx->state = OHS_WRITE_INIT;
-
-    return BIO_printf(rctx->mem, "Content-Length: %ld\r\n\r\n", req_len) > 0
-        && BIO_write(rctx->mem, req, req_len) == (int)req_len;
+    rctx->req = req;
+    if ((req_len = BIO_get_mem_data(req, &req)) <= 0)
```
[the remark this answers is not in the document]
Comment (DDvO) [reply crypto/http/http_client.c]: Yeah, indeed `BIO_ctrl(req, BIO_CTRL_INFO, 0, NULL)` is less confusing.
The nice thing is that it's not strictly required in HTTP to set the content length, 
but we do as far as we can (i.e., as far as supported by the given BIO).

File: include/openssl/http.h (hunk 16 above)
Comment (DDvO) [reply include/openssl/http.h]: Indeed `OSSL_HTTP_DEFAULT_MAX_LINE_LEN` is better - changed.

**Comment by DDvO** [comment]
> Just nits. Otherwise great work!

Nits fixed. Pleased to hear :tada:

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
This will require a rebase after #15251 is fixed. The approval holds assuming the rebase is clean and CI passes.

**Comment by DDvO** [comment]
> This will require a rebase after #15251 is fixed.

Done.

> The approval holds assuming the rebase is clean and CI passes.

This went fine.

**Comment by DDvO** [comment]
Merged - thanks a lot to all to contributed to being able to merge this PR in time!

**Pull request closed without merging**